### PR TITLE
Get and Save Datasets and Datasources to Files

### PR DESCRIPTION
**Pull request by deanp70**
Add functionality to save and get datasets and datasources to files

### CODE REVIEW

File: dagshub/data_engine/client/models.py [hunk 1]
```diff
@@ -89,3 +89,16 @@ class DatasetResult:
...
+@dataclass
+class DatasourceFileStruct:
+    id: Union[str, int]
+    repo: str
+    name: str
+    datasetQuery: Optional[dict]
+
+    @property
+    def has_query(self):
+        return (self.datasetQuery is not None) and (self.datasetQuery["query"] is not None)
```
Comment (kbolashev) [inline dagshub/data_engine/client/models.py:104]: more defensive
```suggestion
        return (self.datasetQuery is not None) and (self.datasetQuery.get("query") is not None)
```

File: dagshub/data_engine/model/datasource.py [hunk 2]
```diff
@@ -529,6 +530,26 @@ def save_dataset(self, name: str):
         self.source.client.save_dataset(self, name)
         log_message(f"Dataset {name} saved")
 
+    def save_to_file(self, path: str = ".", name: str = ""):
```
Comment (kbolashev) [inline dagshub/data_engine/model/datasource.py:533]: Could be useful to return the path to the file as a result of this function, so a Python script can do something else with the resulting file immediately (push it to dagshub, or w/e the user would want)

File: dagshub/data_engine/datasets.py [hunk 3]
```diff
@@ -45,6 +46,19 @@ def get_dataset(repo: str, name: Optional[str] = None, id: Optional[Union[int, s
     return sources[-1]
 
 
+def get_dataset_from_file(path: str):
+    """
+    Load a dataset from a local file
+
+    Args:
+        path: Path to the `.dagshub` file with the relevant dataset
```
Comment (kbolashev) [inline dagshub/data_engine/datasets.py:54]: The sphinx syntax requires double backticks for the code blocks

```suggestion
    Args:
        path: Path to the ``.dagshub`` file with the relevant dataset
```

**Review by kbolashev — APPROVED** [review]
Looks awesome! Thank you!
Adding some nice-to-haves

File: dagshub/data_engine/model/datasource.py (hunk 2 above)
Comment (deanp70) [reply dagshub/data_engine/model/datasource.py]: Done

**Pull request merged**